core/src/com/unciv/logic/automation/civilization/ReligionAutomation.kt

```diff
@@ -193,6 +197,30 @@ object ReligionAutomation {
         cityToBuy.cityConstructions.purchaseConstruction(inquisitorConstruction, -1, true, Stat.Faith)
     }
 
+    private fun buyGreatPerson(civInfo: Civilization) {
+        val greatPersonUnit = civInfo.gameInfo.ruleset.units.values.filter {
+            it.hasUnique(UniqueType.GreatPerson) && !it.hasUnique(UniqueType.MayFoundReligion) //we want to exclude great prophets from the list
+        }
+        val greatPersonConstruction = greatPersonUnit
+            // Get list of cities it can be built in
+            .associateBy({unit -> unit}) { unit -> civInfo.cities.filter { unit.isPurchasable(it.cityConstructions) && unit.canBePurchasedWithStat(it, Stat.Faith) } }
+            .filter { it.value.isNotEmpty() }
+            // And from that list determine the cheapest price
+            .minByOrNull { it.value.minOf { city -> it.key.getStatBuyCost(city, Stat.Faith)!!  }}?.key
+            ?: return
+
+        val validCitiesToBuy = civInfo.cities.filter {
+            (greatPersonConstruction.getStatBuyCost(it, Stat.Faith) ?: return@filter false) <= civInfo.religionManager.storedFaith
+                && greatPersonConstruction.isPurchasable(it.cityConstructions)
+                && greatPersonConstruction.canBePurchasedWithStat(it, Stat.Faith)
+        }
```

Comment:
You should already *have* a list of valid cities from the greatPersonConstruction you did before
Instead of just taking the key, you can take the key (great person) and value (list of cities) and you don't need to check all of this again

The only think missing is the "is less than stored faith" which you can easily add

Reply:
I tried adding your suggestions, but I'm getting type mismatch errors when I try to clean up the code furthers, I think I'll leave it to more experienced devs to improve.